onnxruntime/core/providers/cuda/controlflow/scan.cc

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/controlflow/scan.h"
+
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/tensor/transpose.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace cuda {
+
+template <>
+Scan<8>::Scan(const OpKernelInfo& info) : onnxruntime::Scan<8>(info) {
+  scan::detail::DeviceHelpers helpers;
+
+  helpers.set_data_to_zero_func = [](void* data, size_t size_in_bytes) -> Status {
+    CUDA_RETURN_IF_ERROR(cudaMemset(data, 0, size_in_bytes));
+    return Status::OK();
+  };
+
+  // copy into base class
+  SetDeviceHelpers(helpers);
+}
+
+template <>
+Scan<9>::Scan(const OpKernelInfo& info) : onnxruntime::Scan<9>(info) {
+  scan::detail::DeviceHelpers helpers;
+
+  helpers.transpose_func = [this](const std::vector<size_t>& permutations, const Tensor& input, Tensor& output) {
+    // TODO: We construct a Transpose kernel on each call as doing so is fairly lightweight.
+    // We could potentially keep a single instance and reuse it if that isn't performant enough.
+    const OpKernelInfo& info = OpKernel::Info();
+    return cuda::Transpose::DoTranspose(cuda::Transpose(info), permutations, input, output);
+  };
+
+  // copy into base class
+  SetDeviceHelpers(helpers);
+}
+
+template <>
+Status Scan<8>::Compute(OpKernelContext* ctx) const {
+  // call the base CPU version.
+  // we have this CUDA implementation so the inputs/outputs stay on GPU where possible.
+  // the logic to run the subgraph must be on CPU either way.
+  // technically we don't need this override of Compute, but it will be optimized out and it's easier to debug
+  // that this implementation is being called with it.
+  auto status = onnxruntime::Scan<8>::Compute(ctx);
+  return status;
+}
+
+template <>
+Status Scan<9>::Compute(OpKernelContext* ctx) const {
+  // call the base CPU version.
+  // we have this CUDA implementation so the inputs/outputs stay on GPU where possible.
+  // the logic to run the subgraph must be on CPU either way.
+  // technically we don't need this override of Compute, but it will be optimized out and it's easier to debug
+  // that this implementation is being called with it.
+  auto status = onnxruntime::Scan<9>::Compute(ctx);
+  return status;
+}
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(Scan,
+                                  kOnnxDomain,
+                                  8, 8,
+                                  kCudaExecutionProvider,
+                                  KernelDefBuilder()
+                                      .InputMemoryType<OrtMemTypeCPUInput>(0)  // 'sequence_lens' needs to be on CPU
```

Comment:
>sequence_lens [](start = 83, length = 13)

Putting Scan op in CUDA device has some implications about memory placement for its inputs and outputs. I think a more general way might be to look at the device info in CPU Scan implementation, for the inputs and outputs of Scan node and node placement in its subgraph? Otherwise, Scan9+ may still have suboptimal perf from copies between CPU and GPU.

Reply:
I initially looked at what would be required to check the device info for the Scan inputs/outputs and subgraph inputs/outputs and avoid adding memcpy nodes where possible, but the amount of change to do that is extremely large. We'd have to move around a lot of the initial processing as currently we fully process the main graph followed by subgraphs. Instead you'd have to split major parts of the initialization up a lot differently, and go bottom up so that you have full knowledge of the subgraph partitioning when processing the parent graph. It would also make operator specific logic bleed into general setup as you would need to know which specific inputs of Scan/Loop/If needed to be on CPU and which didn't. There are a few weeks of work in making that sort of change. 

Doing all that would only save you a copy if there was a mismatch between the location of the Scan inputs/outputs and the location of the subgraph inputs/outputs. If we make an assumption that we want minimal gap between the operators available on CUDA and CPU (where possible), that could make such a mismatch rare. 

So whilst it would be more general, I thought the cost/benefit of taking this approach was a clearer win and the changes required are a lot cleaner and better separated. If we still see performance issues with this approach that may validate doing things differently.